src/main/php/lang/ast/emit/PropertyHooks.class.php

```diff
@@ -0,0 +1,171 @@
+<?php namespace lang\ast\emit;
+
+use lang\ast\Code;
+use lang\ast\nodes\{
+  Assignment,
+  Block,
+  InstanceExpression,
+  InvokeExpression,
+  Literal,
+  Method,
+  OffsetExpression,
+  Parameter,
+  ReturnStatement,
+  ScopeExpression,
+  Signature,
+  Variable
+};
+
+/**
+ * Property hooks
+ *
+ * @see  https://wiki.php.net/rfc/property-hooks
+ * @test lang.ast.unittest.emit.PropertyHooksTest
+ */
+trait PropertyHooks {
+
+  protected function rewriteHook($node, $name, $virtual, $literal) {
+
+    // Magic constant referencing property name
+    if ($node instanceof Literal && '__PROPERTY__' === $node->expression) return $literal;
+
+    // Special variable $field, $this->propertyName syntax
+    if ($node instanceof Variable && 'field' === $node->pointer || (
+      $node instanceof InstanceExpression &&
+      $node->expression instanceof Variable && 'this' === $node->expression->pointer &&
+      $node->member instanceof Literal && $name === $node->member->expression
+    )) return $virtual;
+
+    // <T>::$field::hook() => <T>::__<hook>_<field>()
+    if (
+      $node instanceof ScopeExpression &&
+      $node->member instanceof InvokeExpression &&
+      $node->member->expression instanceof Literal &&
+      $node->type instanceof ScopeExpression &&
+      $node->type->member instanceof Variable &&
+      is_string($node->type->type) &&
+      is_string($node->type->member->pointer)
+    ) {
+      return new ScopeExpression($node->type->type, new InvokeExpression(
+        new Literal('__'.$node->member->expression->expression.'_'.$node->type->member->pointer),
+        $node->member->arguments
+      ));
+    }
+
+    foreach ($node->children() as &$child) {
+      $child= $this->rewriteHook($child, $name, $virtual, $literal);
+    }
+    return $node;
+  }
+
+  protected function withScopeCheck($modifiers, $nodes) {
+    if ($modifiers & MODIFIER_PRIVATE) {
+      $check= (
+        '$scope= debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2)[1]["class"] ?? null;'.
+        'if (__CLASS__ !== $scope && \\lang\\VirtualProperty::class !== $scope)'.
+        'throw new \\Error("Cannot access private property ".__CLASS__."::".$name);'
+      );
+    } else if ($modifiers & MODIFIER_PROTECTED) {
+      $check= (
+        '$scope= debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2)[1]["class"] ?? null;'.
+        'if (__CLASS__ !== $scope && !is_subclass_of($scope, __CLASS__) && \\lang\\VirtualProperty::class !== $scope)'.
+        'throw new \\Error("Cannot access protected property ".__CLASS__."::".$name);'
+      );
+    } else if (1 === sizeof($nodes)) {
+      return $nodes[0];
+    } else {
+      return new Block($nodes);
+    }
+
+    array_unshift($nodes, new Code($check));
```

Comment:
This can be written as `return new Block([new Code($check), ...$nodes]);` as of PHP 7.4.0, see https://wiki.php.net/rfc/spread_operator_for_array and https://github.com/xp-framework/rfc/issues/343